File: 2019/day8.js
/*

--- Day 8: Space Image Format ---

The Elves' spirits are lifted when they realize you have an opportunity to reboot one of their Mars rovers, and so they are curious if you would spend a brief sojourn on Mars. You land your ship near the rover.

When you reach the rover, you discover that it's already in the process of rebooting! It's just waiting for someone to enter a BIOS password. The Elf responsible for the rover takes a picture of the password (your puzzle input) and sends it to you via the Digital Sending Network.

Unfortunately, images sent via the Digital Sending Network aren't encoded with any normal encoding; instead, they're encoded in a special Space Image Format. None of the Elves seem to remember why this is the case. They send you the instructions to decode it.

Images are sent as a series of digits that each represent the color of a single pixel. The digits fill each row of the image left-to-right, then move downward to the next row, filling rows top-to-bottom until every pixel of the image is filled.

Each image actually consists of a series of identically-sized layers that are filled in this way. So, the first digit corresponds to the top-left pixel of the first layer, the second digit corresponds to the pixel to the right of that on the same layer, and so on until the last digit, which corresponds to the bottom-right pixel of the last layer.

For example, given an image 3 pixels wide and 2 pixels tall, the image data 123456789012 corresponds to the following image layers:

Layer 1: 123
         456

Layer 2: 789
         012
         
The image you received is 25 pixels wide and 6 pixels tall.

To make sure the image wasn't corrupted during transmission, the Elves would like you to find the layer that contains the fewest 0 digits. On that layer, what is the number of 1 digits multiplied by the number of 2 digits?


--- Part Two ---

Now you're ready to decode the image. The image is rendered by stacking the layers and aligning the pixels with the same positions in each layer. The digits indicate the color of the corresponding pixel: 0 is black, 1 is white, and 2 is transparent.

The layers are rendered with the first layer in front and the last layer in back. So, if a given position has a transparent pixel in the first and second layers, a black pixel in the third layer, and a white pixel in the fourth layer, the final image would have a black pixel at that position.

For example, given an image 2 pixels wide and 2 pixels tall, the image data 0222112222120000 corresponds to the following image layers:

Layer 1: 02
         22

Layer 2: 11
         22

Layer 3: 22
         12

Layer 4: 00
         00

Then, the full image can be found by determining the top visible pixel in each position:

The top-left pixel is black because the top layer is 0.
The top-right pixel is white because the top layer is 2 (transparent), but the second layer is 1.
The bottom-left pixel is white because the top two layers are 2, but the third layer is 1.
The bottom-right pixel is black because the only visible pixel in that position is 0 (from layer 4).

So, the final image looks like this:

01
10

What message is produced after decoding your image?

*/

// OPTIONAL VARIABLES
const DISPLAY_EXTRA_INFO = true;

function layerImageData (part, inputStr, extraParam, DEBUG = false) {
  
  const [ W, H ] = extraParam;
  const SIZE = W * H;

  // ANALYZE
  if (part === 1) {

    const LAYER_DATA = [];

    for (let i = 0; i < inputStr.length; i += SIZE) {

      const layer = inputStr.slice(i, i + SIZE);
      let num0 = 0;
      let num1 = 0;
      let num2 = 0;
      for (let c of layer) {
        if (c === '0') ++num0;
        if (c === '1') ++num1;
        if (c === '2') ++num2;
      }
      LAYER_DATA.push({ num0, num1, num2 });

    }

    let num0 = SIZE;
    let layerWithFewest0 = null;
    for (const layer of LAYER_DATA) {
      if (layer.num0 < num0) {
        num0 = layer.num0;
        layerWithFewest0 = layer;
      }
    }

    return layerWithFewest0.num1 * layerWithFewest0.num2;

  } else {

    const FINAL_IMAGE = Array.from({length: H}, () => Array(W).fill(null));

    for (let i = 0; i < inputStr.length; i += SIZE) {

      const layer = inputStr.slice(i, i + SIZE);

      for (let j = 0; j < layer.length; j += W) {

        const row = Math.floor(j / W);
        const rowData = layer.slice(j, j + W);

        for (let col = 0; col < W; ++col) {

          if (FINAL_IMAGE[row][col] === null && rowData[col] !== '2') {
            FINAL_IMAGE[row][col] = rowData[col];
          }

        }

      }

    }

    console.log(FINAL_IMAGE.map(row => row.map(c => c === '0' ? '   ' : ' . ').join('')).join('\n'));
    return FINAL_IMAGE.map(row => row.join('')).join('\n');

  }
}

// TEST CASES

const test = require('./_test');
const testNum = [1];
let input, expected;
const func = layerImageData;
const sortedFunc = (...args) => func(...args).sort();                   // used when the order of the output does not matter
const modFunc = (...args) => func(...args) % 1000000007;                // used when the output is very large
const skippedTests = new Set([  ]);
const lowestTest = 0;
const highestTest = 0;

const fs = require('fs');
const path = require('path');
const DAY_NUM = __filename.split('.js')[0].split('day')[1];
const INPUT_PATH = path.join(__dirname, `day${DAY_NUM}-input.txt`);
const actualInput = fs.readFileSync(INPUT_PATH, 'utf8');
const parseSampleInput = s => s.split('').map(c => c === '\n' ? '\r\n' : c).join('');

const sampleInput = parseSampleInput(
`123456789012`
);

const sampleInput2 = parseSampleInput(
`0222112222120000`
);

// Test case 1
input = {
  part: 1,
  inputStr: sampleInput,
  extraParam: [ 3, 2 ],
  DEBUG: true,
};
expected = 1;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 2
input = {
  part: 1,
  inputStr: actualInput,
  extraParam: [ 25, 6 ],
};
expected = 1560;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 3
input = {
  part: 2,
  inputStr: sampleInput2,
  extraParam: [ 2, 2 ],
  DEBUG: true,
};
expected =
`01
10`
;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 4
input = {
  part: 2,
  inputStr: actualInput,
  extraParam: [ 25, 6 ],
};
expected =
`1001001100011001001010010
1001010010100101001010010
1001010000100001001011110
1001010110100001001010010
1001010010100101001010010
0110001110011000110010010`
;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

File: 2023/day8.py
"""

--- Day 8: Haunted Wasteland ---

You're still riding a camel across Desert Island when you spot a sandstorm quickly approaching. When you turn to warn the Elf, she disappears before your eyes! To be fair, she had just finished warning you about ghosts a few minutes ago.

One of the camel's pouches is labeled "maps" - sure enough, it's full of documents (your puzzle input) about how to navigate the desert. At least, you're pretty sure that's what they are; one of the documents contains a list of left/right instructions, and the rest of the documents seem to describe some kind of network of labeled nodes.

It seems like you're meant to use the left/right instructions to navigate the network. Perhaps if you have the camel follow the same instructions, you can escape the haunted wasteland!

After examining the maps for a bit, two nodes stick out: AAA and ZZZ. You feel like AAA is where you are now, and you have to follow the left/right instructions until you reach ZZZ.

This format defines each node of the network individually. For example:

RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)

Starting with AAA, you need to look up the next element based on the next left/right instruction in your input. In this example, start with AAA and go right (R) by choosing the right element of AAA, CCC. Then, L means to choose the left element of CCC, ZZZ. By following the left/right instructions, you reach ZZZ in 2 steps.

Of course, you might not find ZZZ right away. If you run out of left/right instructions, repeat the whole sequence of instructions as necessary: RL really means RLRLRLRLRLRLRLRL... and so on. For example, here is a situation that takes 6 steps to reach ZZZ:

LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)

Starting at AAA, follow the left/right instructions. How many steps are required to reach ZZZ?


--- Part Two ---

The sandstorm is upon you and you aren't any closer to escaping the wasteland. You had the camel follow the instructions, but you've barely left your starting position. It's going to take significantly more steps to escape!

What if the map isn't for people - what if the map is for ghosts? Are ghosts even bound by the laws of spacetime? Only one way to find out.

After examining the maps a bit longer, your attention is drawn to a curious fact: the number of nodes with names ending in A is equal to the number ending in Z! If you were a ghost, you'd probably just start at every node that ends with A and follow all of the paths at the same time until they all simultaneously end up at nodes that end with Z.

For example:

LR

11A = (11B, XXX)
11B = (XXX, 11Z)
11Z = (11B, XXX)
22A = (22B, XXX)
22B = (22C, 22C)
22C = (22Z, 22Z)
22Z = (22B, 22B)
XXX = (XXX, XXX)

Here, there are two starting nodes, 11A and 22A (because they both end with A). As you follow each left/right instruction, use that instruction to simultaneously navigate away from both nodes you're currently on. Repeat this process until all of the nodes you're currently on end with Z. (If only some of the nodes you're on end with Z, they act like any other node and you continue as normal.) In this example, you would proceed as follows:

    Step 0: You are at 11A and 22A.
    Step 1: You choose all of the left paths, leading you to 11B and 22B.
    Step 2: You choose all of the right paths, leading you to 11Z and 22C.
    Step 3: You choose all of the left paths, leading you to 11B and 22Z.
    Step 4: You choose all of the right paths, leading you to 11Z and 22B.
    Step 5: You choose all of the left paths, leading you to 11B and 22C.
    Step 6: You choose all of the right paths, leading you to 11Z and 22Z.

So, in this example, you end up entirely on nodes that end in Z after 6 steps.

Simultaneously start on every node that ends with A. How many steps does it take before you're only on nodes that end with Z?


"""

# LIBRARIES
import time
import pathlib
import os
from math import *
from functools import *
from sys import *
from collections import *
from queue import *

# MODULES
from _test import test

# OPTIONAL VARIABLES
DISPLAY_EXTRA_INFO = True
# DISPLAY_EXTRA_INFO = False

def find_sync_point_of_all_cycles(part, input_str, DEBUG = False):

  # DATA STRUCTURE

  DATA = {}


  # PARSE INPUT DATA

  [ INSTRUCTIONS, DATA_STR ] = input_str.split('\n\n')  
  for line in DATA_STR.split('\n'):
    split = line.split(' = ')
    key = split[0]
    left_and_right_split = split[1].split(', ')
    left = left_and_right_split[0][1:]
    right = left_and_right_split[1][:-1]
    DATA[key] = { 'L': left, 'R': right }


  # ANALYZE

  LIMIT = maxsize

  if part == 1:                                   # PART 1: GO FROM 'AAA' TO 'ZZZ'

    steps = 0
    idx = 0
    node = 'AAA'
    while steps < LIMIT:
      if node == 'ZZZ': return steps
      node = DATA[node][INSTRUCTIONS[idx]]
      idx = (idx + 1) % len(INSTRUCTIONS)
      steps += 1

    print('STEPS EXCEEDED LIMIT')
    assert(False)

  else:                                           # PART 2: SEND MULTIPLE THREADS ALONG ALL '*A' TO '*Z' AND FIND POINT IN TIME WHERE CYCLES MATCH

    # find all start nodes

    node_list = []
    for line in DATA_STR.split('\n'):
      split = line.split(' = ')
      key = split[0]
      if key[-1] == 'A':
        node_list.append(key)


    # for each node, find period of repeat

    periods = {}

    for node in node_list:
      periods[node] = {}
      steps = 0
      idx = 0
      curr_node = node
      while steps < LIMIT:
        if curr_node[-1] == 'Z':
          periods[node][curr_node] = steps
        if (curr_node in periods[node]):

          # NOTE: IT SEEMS THAT, COINCIDENTALLY (I THINK?) BOTH IN THE SAMPLE DATA AND IN THE REAL DATA,
          # THE NUMBER OF STEPS IT TAKES TO GO FROM A TO THE FIRST Z NODE IS EQUAL TO THE NUMBER OF STEPS
          # IT TAKES TO GO FROM THE FIRST Z NODE BACK TO ITSELF. (ALSO, EACH A NODE UNIQUELY VISITS ONLY
          # ONE UNIQUE Z NODE.) IF THESE THINGS WERE NOT TRUE, THIS SOLUTION WOULD NEED TO BE TWEAKED.
          # HOWEVER, I HAVE INSPECTED THE DATA WITH THE FOLLOWING AND CONFIRMED THIS TO BE THE CASE.
          if DISPLAY_EXTRA_INFO:
            print(f'ANALYZING {node}:')
            print(f'Total steps on first Z node to repeat: {steps}')
            print(f'Steps from A to first Z: {periods[node][curr_node]}')
            print('---')
            
          break
        curr_node = DATA[curr_node][INSTRUCTIONS[idx]]
        idx = (idx + 1) % len(INSTRUCTIONS)
        steps += 1
      
      if steps == LIMIT:
        print('STEPS EXCEEDED LIMIT')
        assert(False)
    
    if DISPLAY_EXTRA_INFO: print(periods)

    # # UTILITY FUNCTIONS - ALTHOUGH WE DON'T NEED THIS BECAUSE math ALREADY HAS lcm FUNCTION!

    # # uses Euclidean algorithm (https://en.wikipedia.org/wiki/Euclidean_algorithm)
    # # credit to Phrogz (https://stackoverflow.com/questions/4652468/is-there-a-javascript-function-that-reduces-a-fraction)
    # def GCD(num, denom):
    #   num = abs(num)
    #   denom = abs(denom)
    #   return GCD(denom, num % denom) if denom else num

    # # credit to w3resource (https://www.w3resource.com/javascript-exercises/javascript-math-exercise-10.php)
    # def LCM(num1, num2):
    #   return 0 if (not num1 or not num2) else abs((num1 * num2) // GCD(num1, num2))


    # FIND LCM OF ALL PERIODS

    steps_list = [ list(periods[start_node].values())[0] for start_node in periods ]
    if DISPLAY_EXTRA_INFO:
      for i in range(len(node_list)):
        print(f"{node_list[i]}: {steps_list[i]} steps")

    return reduce(lambda x, y: lcm(x, y), steps_list)


# TEST CASES

test_num = [1]
test_input = None
test_expected = None
func = find_sync_point_of_all_cycles
skipped_tests = set([ 2, 3, 4 ])
skipped_tests = set([ 3, 4 ])
skipped_tests = set([ 4 ])
skipped_tests = set([  ])
lowest_test = 0
highest_test = 0

FILENAME = pathlib.Path(__file__).name
FILENAME_SPLIT = FILENAME.split('.py')[0].split('day')
DAY_NUM = FILENAME_SPLIT[1] if len(FILENAME_SPLIT) > 1 else None
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_PATH = CURRENT_DIR + '\\' + (f"day{DAY_NUM}-input.txt" if DAY_NUM else 'template-input.txt')
actual_input = open(INPUT_PATH, 'r').read()

sample_input = """RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)"""

sample_input2 = """LR

11A = (11B, XXX)
11B = (XXX, 11Z)
11Z = (11B, XXX)
22A = (22B, XXX)
22B = (22C, 22C)
22C = (22Z, 22Z)
22Z = (22B, 22B)
XXX = (XXX, XXX)"""

# Test case 1
test_input = {
  'part': 1,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 2
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 2
test_input = {
  'part': 1,
  'input_str': actual_input,
}
test_expected = 17263
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 3
test_input = {
  'part': 2,
  'input_str': sample_input2,
  'DEBUG': True,
}
test_expected = 6
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 4
test_input = {
  'part': 2,
  'input_str': actual_input,
}
test_expected = 14631604759649
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)